telethon/network/connection/connection.py

```diff
@@ -46,58 +59,76 @@ def __init__(self, ip, port, dc_id, *, loggers, proxy=None, local_addr=None):
-                timeout=timeout
-            )
+        if not self._proxy:
+            connect_coroutine = asyncio.open_connection(
+                host=self._ip,
+                port=self._port,
+                ssl=ssl,
+                local_addr=local_addr)
         else:
-            import socks
-            if ':' in self._ip:
-                mode, address = socket.AF_INET6, (self._ip, self._port, 0, 0)
-            else:
-                mode, address = socket.AF_INET, (self._ip, self._port)
 
-            s = socks.socksocket(mode, socket.SOCK_STREAM)
-            if isinstance(self._proxy, dict):
-                s.set_proxy(**self._proxy)
+            if isinstance(self._proxy, (tuple, list)):
+                proxy, auth, remote_resolve = self._parse_proxy(*self._proxy)
+            elif isinstance(self._proxy, dict):
+                proxy, auth, remote_resolve = self._parse_proxy(**self._proxy)
             else:
-                s.set_proxy(*self._proxy)
-
-            s.settimeout(timeout)
-            if self._local_addr is not None:
-                s.bind((self._local_addr, None))
-            await asyncio.wait_for(
-                asyncio.get_event_loop().sock_connect(s, address),
-                timeout=timeout
-            )
-            if ssl:
-                if ssl_mod is None:
-                    raise RuntimeError(
-                        'Cannot use proxy that requires SSL'
-                        'without the SSL module being available'
-                    )
-
-                s = ssl_mod.wrap_socket(
-                    s,
-                    do_handshake_on_connect=True,
-                    ssl_version=ssl_mod.PROTOCOL_SSLv23,
-                    ciphers='ADH-AES256-SHA'
-                )
-                
-            s.setblocking(False)
-
-            self._reader, self._writer = await asyncio.open_connection(sock=s)
+                raise ValueError("Unknown proxy format: {}".format(self._proxy.__class__.__name__))
 
+            connect_coroutine = aiosocks.open_connection(
+                proxy=proxy,
+                proxy_auth=auth,
+                dst=(self._ip, self._port),
+                remote_resolve=remote_resolve,
+                ssl=ssl,
+                local_addr=local_addr)
+
+        self._reader, self._writer = await asyncio.wait_for(connect_coroutine, timeout=timeout)
         self._codec = self.packet_codec(self)
         self._init_conn()
         await self._writer.drain()
 
+    @staticmethod
+    def _parse_proxy(proxy_type, addr, port, rdns=True, username=None, password=None):
+
+        proxy, auth = None, None
+
+        if isinstance(proxy_type, str):
+            proxy_type = proxy_type.lower()
+
+        # We do the check for numerical values here
+        # to be backwards compatible with PySocks proxy format,
+        # (since socks.SOCKS5 = 2 and socks.SOCKS4 = 1)
```

Comment:
There is also a type `3` for HTTP:

https://github.com/Anorov/PySocks/blob/c2fa43cbe1091e799e248e8e4433978916791a8b/socks.py#L30

Does AioSocks have something similar?

Reply:
@Lonami 
I don't think so. AioSocks supports only SOCKS4 and SOCKS5.

Reply:
@Lonami 

Maybe we can try to connect as SOCKS5 if `socks.HTTP == 3` provided?

Reply:
@Lonami 

I also have a version with `python-socks` [(link)](https://github.com/romis2012/python-socks) instead of `aiosocks`.
It also works good and supports HTTP (tunneling). 

Maybe we can go with it?

Reply:
If someone relied on HTTP proxies and we just removed that, then it would break their workflow. That library claims to require Python >= 3.6, and in theory Telethon still supports Python 3.5, so that would also be a breaking change…

Our options are:
* fall back to blocking PySocks again for HTTP proxies.
* look for another async proxy library.
* send a pull request to python-socks to add support for Python 3.5 and hope they merge it.
  * alternatively keep the fork and point users to install that instead.

Reply:
@Lonami 

I propose to go with: `python-socks` for Python >= 3,6 and leave `PySocks` for Python <= 3.5.
The reasons for this are:
* `aiosocks` does not support `HTTP`, `python-socks` does
* `aiosocks` was not updated for a long time, maybe not maintained by author anymore
* After some testing I have found a bug in `aiosocks` (see [this](https://github.com/nibrag/aiosocks/issues/43)),
which is internal to library and cannot be fixed without patching lib itself (related to `loop.call_soon` and `asyncio.wait_for` behaviour)
* `python-socks` has support for `curio` and `trio` in case if `Telethon` is going to add support for those later.

What do you think? If it is ok, I can close this pull request and add new one with `python-socks`.

Reply:
There are no near-term plans to support anything but `asyncio` in Telethon, so if that were to happen it would most likely not be in v1. The other reasons are good though; I think proper support in Python 3.6 and leaving the buggy behaviour (which isn't terrible, just not ideal) in 3.5 is a good approach.

Let's use python-socks for Python 3.6+ and keep the old PySocks code for Python 3.5. Feel free to send a new pull request if you prefer, or just channge this one, I'm going to squash the commits after we're done anyway.